src/ppl/src/IOPlacer.cpp

```diff
@@ -1008,7 +1013,33 @@ void IOPlacer::initConstraints()
   std::reverse(constraints_.begin(), constraints_.end());
   for (Constraint &constraint : constraints_) {
     getPinsFromDirectionConstraint(constraint);
+    createSectionsPerConstraint(constraint);
+    int num_slots = 0;
+    for (Section sec : constraint.sections) {
+      num_slots += sec.num_slots;
+    }
+    constraint.pins_per_slots = (float)constraint.pin_list.size()/num_slots;
   }
+  sortConstraints();
+}
+
+void IOPlacer::sortConstraints() {
+  std::stable_sort(constraints_.begin(),
+                   constraints_.end(),
+                   [&](const Constraint& c1, const Constraint& c2) {
+                     return (c1.pins_per_slots < c2.pins_per_slots) && overlappingConstraints(c1, c2);
```

Comment:
non-overlapping constraints will not be weakly ordered ( c1 < c2 and c2 < c1 will both be false).

Reply:
My intention is to keep the ordering defined by the user whenever is possible. I want to sort only the overlapping constraints to avoid failures due to not enough free slots.

Reply:
So in effect treating everything non-overlapping as equal so that stable_sort will keep the user order.  Clever but worth a comment as it is non-obvious.